### PR TITLE
:tada: etl-metamigrate CLI tool for generating grapher YAML configs

### PR DESCRIPTION
**Pull request by Marigold**
`etl-metadata-migrate` is a CLI tool to help with **migrating grapher pages to data pages**. We have [a list](https://docs.google.com/spreadsheets/d/1SyyRR_iaokoc-tkDVQ24F4gW7cf47mLRdEyr8ZSBYo0/edit#gid=0) of the most important metrics that should get their own data page, which means creating YAML file with the new metadata structure and reasonable grapher config.

This tool tries to automate as much as possible by creating or updating YAML file in the grapher channel. It adds **all existing metadata about the indicator** to the YAML file to make it self-contained. It won't change the metadata structure for you, but at least give you a template to reduce the pain.

## How to review this

Pick one of indicators from the [migration list](https://docs.google.com/spreadsheets/d/1SyyRR_iaokoc-tkDVQ24F4gW7cf47mLRdEyr8ZSBYo0/edit#gid=0), run e.g.
```
STAGING=mojmir etl-metadata-migrate --chart-slug life-expectancy
```
and then follow the instructions. Is it helpful? Do you have any ideas for improvements?

## AI Automation

SInce all existing metadata is in the YAML file and fields that need to be filled are prefixed with `TBD`, it should be theoretically possible to automate metadata restructuring with ChatGPT API (this is independent on `etl-metadata-migrate`, the AI function would just take YAML file and return YAML). We could instruct it with [our docs](https://docs.owid.io/projects/etl/architecture/metadata/reference/origin/) and let it will the missing pieces.

Having this would let us scale data pages migration. It could even mean that I can do all of them myself and just let others review them instead of splitting the work into 6 people.

@veronikasamborska1994 @lucasrodes what do you think? You are our ChatGPT experts.

### CODE REVIEW

**Comment by Marigold** [comment]
Issues when migrating [CO2 emissions](https://ourworldindata.org/grapher/consumption-co2-per-capita) data page:
* Tags [CO2 and greenhouse emissions](https://ourworldindata.org/co2-and-greenhouse-gas-emissions) and [Climate change](https://ourworldindata.org/explorers/climate-change) don't exist in `tags` table
* There are two tags `Energy` in `tags` table and both are shown in ETL data page

File: etl/steps/data/grapher/happiness/2023-03-20/happiness.meta.yml [hunk 1]
```diff
@@ -0,0 +1,171 @@
+dataset:
```
Comment (Marigold) [inline etl/steps/data/grapher/happiness/2023-03-20/happiness.meta.yml:1]: @pabloarosado can I please ask you to check the **structure** of this migrated data page YAML before I delegate it to their owners? I'm looking for structural issues that I could correct in all migrated data pages, not for specific metadata issues (these will be up to owners to fix).

Here's [ETL-based data page](http://staging-site-mojmir/admin/datapage-preview/577009) and here's [JSON based](https://ourworldindata.org/grapher/happiness-cantril-ladder). Thank you!

File: etl/steps/data/grapher/democracy/2023-03-02/vdem.meta.yml [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+              gdoc_id: 1gGburArxglFdHXeTLotFW4TOOLoeRq5XW6UfAdKtaAw
+            - fragment_id: vdem-scoring
+              gdoc_id: 1gGburArxglFdHXeTLotFW4TOOLoeRq5XW6UfAdKtaAw
+            - fragment_id: vdem-coverage
+              gdoc_id: 1gGburArxglFdHXeTLotFW4TOOLoeRq5XW6UfAdKtaAw
+            - fragment_id: vdem-measured
+              gdoc_id: 1gGburArxglFdHXeTLotFW4TOOLoeRq5XW6UfAdKtaAw
+            - fragment_id: vdem-more-info
+              gdoc_id: 1gGburArxglFdHXeTLotFW4TOOLoeRq5XW6UfAdKtaAw
+          grapher_config:
```
Comment (Marigold) [inline etl/steps/data/grapher/democracy/2023-03-02/vdem.meta.yml:118]: @danyx23 I used the **Copy as YAML** button from grapher and pasted grapher_config as it is to data pages. Can I please ask you to check a couple of these if any fields are obviously wrong or useless and don't belong to the data page YAML?

I'm unsure about fields like `slug`, `$schema` or `relatedQuestions`. If we find such fields we should exclude them from **Copy as YAML** and show linting error.

File: etl/steps/data/grapher/happiness/2023-03-20/happiness.meta.yml (hunk 1 above)
Comment (pabloarosado) [reply etl/steps/data/grapher/happiness/2023-03-20/happiness.meta.yml:1]: This one looks good, do you want me to check the others?

Comment (Marigold) [reply etl/steps/data/grapher/happiness/2023-03-20/happiness.meta.yml:1]: Nope, thanks a lot!

**Comment by veronikasamborska1994** [comment]
hey @Marigold! this is awesome and exactly what I was wondering if we'd be able to do when I first heard about the metadata format changes! thanks so much for making this happen! :) 

I think it should definitely be possible to just feed the yaml files to the OpenAI API and ask it to do a combination of inference + bing search to fill out some of these variables. We should be able to do this for any yaml file we generate in general, and maybe not even just for the ones where we already have some pre-existing info. I'll have some time next week to do some experimenting on this! 

As a side note, for this particular function I think we can automatically fill out description_short from subtitle if exists? 

p.s. I am also getting an error when I try to run it for one of the [key charts](https://ourworldindata.org/grapher/cross-country-literacy-rates?tab=map) I am responsible for making the data page for. When I run `STAGING=veronika etl-metadata-migrate --chart-slug cross-country-literacy-rates` I am getting this error ` assert variable.catalogPath, f"Variable {var_id} does not come from ETL. Migrate it there first.` It's definitely already on the etl but the variable relies on a [dataset](https://github.com/owid/etl/tree/master/snapshots/education/2018-04-18) which you helped me migrate.

**Comment by veronikasamborska1994** [comment]
p.s. I've been using ChatGPT to do the conversions in the snapshot step which work super well - 

this is what I got for one my datasets 
```

# Learn more at:
# http://docs.owid.io/projects/etl/architecture/metadata/reference/origin/
meta:
  origin:
    # Data product / Snapshot
    title: Public Expenditure on Education OECD - Tanzi & Schuknecht (2000)
    description: >
      This dataset is sourced from Tanzi & Schuknecht's 2000 publication and 
      includes data from various historical sources such as the League of Nations 
      Statistical Yearbook, Mitchell (1962), OECD Education at a Glance (1996), 
      UNESCO World Education Report (1993), UNDP Human Development Report (1996), 
      and UN World Economics Survey. There are noted discrepancies in the data, 
      particularly when newer sources are compared with the values reported by 
      Tanzi & Schuknecht, due to the lack of clear source prioritization for each 
      year and country.
    date_published: '2000'
    # Citation
    producer: Vito Tanzi and Ludger Schuknecht
    citation_full: 'Tanzi, Vito, and Ludger Schuknecht. Public Spending in the 20th 
      Century: A Global Perspective. 2000.'
    # Files
    url_main: https://link.springer.com/article/10.1023%2FA%3A1017578302202?LI=true
    date_accessed: 2017-09-30
    # License
    license:
      name: Not specified
      url: Not specified
wdir: ../../../data/snapshots/education/2017-09-30
outs:
- md5: 70215292759c49a6bdb664321f757c15
  size: 3666
  path: public_expenditure.feather
```

and this was the original snapshot 
```

meta:
  source:
    name: Tanzi & Schuknecht (2000)
    description: 'The underlying sources for Tanzi & Schuknecht (2000) include: League
      of Nations Statistical Yearbook (various years), Mitchell (1962), OECD Education
      at a Glance (1996), UNESCO World Education Report (1993), UNDP Human Development
      Report (1996), UN World Economics Survey (various years). To the extent that
      the authors do not specify which sources were prioritised for each year/country,
      it is not possible for us to reliably extend the time series with newer data.
      For instance, the OECD Education at a Glance report (1998), which presents estimates
      for the years 1990 and 1995, suggests discrepancies with the values reported
      by Tanzi & Schuknecht (2000) for 1993.'
    url: https://link.springer.com/article/10.1023%2FA%3A1017578302202?LI=true
    date_accessed: 2017-09-30
    published_by: Vito and Schuknecht (2000) Public Spending in the 20th Century A
      Global Perspective
  name: Public Expenditure on Education OECD - Tanzi & Schuknecht (2000)
  description: ''
wdir: ../../../data/snapshots/education/2017-09-30
outs:
- md5: 70215292759c49a6bdb664321f757c15
  size: 3666
  path: public_expenditure.feather

```

pretty good I think!

**Comment by Marigold** [comment]
> p.s. I am also getting an error when I try to run it for one of the [key charts](https://ourworldindata.org/grapher/cross-country-literacy-rates?tab=map) I am responsible for making the data page for. When I run STAGING=veronika etl-metadata-migrate --chart-slug cross-country-literacy-rates I am getting this error  assert variable.catalogPath, f"Variable {var_id} does not come from ETL. Migrate it there first.

Your staging server is quite outdated and the chart still points to the old dataset. You should either run
```
ssh owid@staging-site-veronika 'cd owid-grapher && make refresh'
```
which will sync your staging MySQL with production or kill your server with `owid-lxc destroy staging-site-veronika` and then rebase & commit in grapher (the latter might be more robust).

> As a side note, for this particular function I think we can automatically fill out description_short from subtitle if exists?

Thanks for the tip, done!

> p.s. I've been using ChatGPT to do the conversions in the snapshot step which work super well -

Looks great! If we make it work like this, then we get almost automatic data page migration. This would save so much time (and nerves...).

**Review by lucasrodes — APPROVED** [review]
This is very neat! Thanks for doing this, @Marigold!

Some thoughts for minor improvements:

- We could print in the terminal when the new grapher YAML file is created so the user can see its path, open it and edit it further. Maybe in the instructions list printed after calling `etl-metadata-migrate`, e.g. "Go to path/to/yaml and edit it accordingly".
- You've added placeholders (e.g. "TBD - major or minor") in the Grapher YAML file. I think it is good that something is rendered in the data page review. In addition, we should somehow prevent the user from being able to execute the Grapher step until properly edited (I can see how we accidentally pushed these templates to live). Can we introduce an error on the YAML file on purpose?
- Maybe in the future, we could have an option like `etl-metadata-migrate --show-pending-slugs` or similar, which shows the list of pending charts to be migrated (maybe only those single-indicator, referenced, etc.)

**Comment by Marigold** [comment]
@lucasrodes thanks for feedback, I've added 1. and keeping 2. and 3. as future improvements (it's not particularly urgent, but I agree it'd be very nice to have).

**Pull request merged**